node/src/container_chain_spawner.rs

```diff
@@ -330,14 +416,24 @@ impl ContainerChainSpawner {
             sc_service::error::Result::Ok(())
         };
 
-        async {
+        async move {
             match try_closure().await {
                 Ok(()) => {}
+                Err(sc_service::error::Error::Application(e)) if e.is::<NeedsRestart>() => {
+                    let e = e.downcast::<NeedsRestart>().unwrap();
+
+                    log::info!("Restarting container chain {}", container_chain_para_id);
+                    // self.spawn must return a boxed future because of the recursion here
+                    e.self2
+                        .spawn(container_chain_para_id, start_collation, e.warp_sync)
+                        .await;
```

Comment:
Unfortunately this doesn't work because the database doesn't have time to close, resulting in a db lock error. We should add a sleep(10) here, or check if the db is still alive using the container_chain_monitor.